contracts/Removal.sol

```diff
@@ -351,6 +288,29 @@ contract Removal is
     return super.supportsInterface(interfaceId);
   }
 
+  function isApprovedForAll(address account, address operator)
+    public
+    view
+    override
+    returns (bool)
+  {
+    return
+      account == address(_market) ||
```

Comment:
confused about this because this will always return true if asking whether someone is an approved operator for the market, which doesn't seem right at all, especially because in `setApprovalForAll` the market is explicitly not allowed to become an operator for anything.  maybe I'm missing something

Reply:
the intent of this is to hijack the internals of `safe(Batch)TransferFrom` such that we don't ever need to set an approval for the market and that the market is an approved sender for all accounts always (such that it enables calling the release and withdraw functions). Approvals don't currently have a use case otherwise, but they aren't precluded in case we want to leverage it later. The only reason setting the market as approved/unapproved reverts is because setting the approval there wouldn't do anything given the way this check works.

Previously we were setting approvals on every mint, but this was adding to the overhead of minting as you either had to set state or load state on every mint. This circumvents this and puts it further down the stack as it is always only checked if the sender isn't the owner  e.g., `from == _msgSender() || isApprovedForAll(from, _msgSender()),`

Does that make sense? 

I suppose another way to do this would be to approve the market in the initializer and then continuing to disallow changing that approval in the approval function... Maybe that would be preferable?

Reply:
I think I'd include that in a follow up PR though unless this is creating unexpected behavior that I'm otherwise missing

Reply:
here's how this is used internally

```sol
 function safeTransferFrom(
        address from,
        address to,
        uint256 id,
        uint256 amount,
        bytes memory data
    ) public virtual override {
        require(
            from == _msgSender() || isApprovedForAll(from, _msgSender()),
            "ERC1155: caller is not token owner nor approved"
        );
        _safeTransferFrom(from, to, id, amount, data);
    }
```

AFAICT unless you're the market (_msgSender + from) this should be safe-- but maybe I'm overlooking something obvious

Reply:
```
ontract Market_withdraw_reverts is MarketBalanceTestHelper {
  function setUp() external {
    _removalIds = _seedRemovals({
      to: _namedAccounts.supplier,
      count: 1,
      list: true
    });
    _suppliers = new address[](1).fill(_namedAccounts.supplier);
    _expectedRemovalBalances = [0];
    _expectedMarketSupply = _amountPerRemoval * _removalIds.length;
    _expectedTokenCount.set(_namedAccounts.supplier, 0);
    _expectedTokenCount.set(address(_market), _removalIds.length);
    _assertCorrectStates();
  }

  function test_correctlyReverts() external {
    vm.prank(_namedAccounts.supplier2);
    vm.expectRevert(UnauthorizedWithdrawal.selector);
    _market.withdraw(_removalIds[0]);
    _assertCorrectStates();
  }

  function test_incorrectlyAllow() external {
    vm.prank(_namedAccounts.supplier2);
    vm.expectRevert(UnauthorizedWithdrawal.selector);
    _removal.safeTransferFrom({
      from: address(_market),
      to: _namedAccounts.supplier2,
      id: _removalIds[0],
      amount: 1 ether,
      data: ""
    });
    _assertCorrectStates();
  }
}
```
**`[FAIL. Reason: Call did not revert as expected] test_incorrectlyAllow`**
you're right, this allows be to withdraw by `safeTransferFrom` on the removal. good catch. I guess the initializer is the best approach here

Reply:
innnnnnteresting... I actually didn't catch this issue, was mostly just confused, but I see now what's happening.  I wonder if that's because we really just wanted to be asking (inside `isApprovedForAll`) to `return operator == address(_market) || ....` instead of `return account == address(_market) || ....` as it's written.  we want to know if it's the market _calling_ `safe(Batch)TransferFrom`, not whether it's the market from which the transfer is being pulled

Reply:
I think a piece that's still not making sense is actually why the market needs to be an approved operator at all... I know that we were setting nori admin address (whoever deploys i think) to be an operator so that we can list supplier removals for sale in the market, but once that happens the market literally owns the removals anyway, and so is free to transfer them to the certificate.  and `release` is initiated by the removal contract (who also performs the burning of the market's balance because the message sender for that call is a nori admin address), and `withdraw` has its own set of permissions around it and again just involves the market transferring its own balance of a removal to someone else.  so I still can't understand why the market contract needs permanent operator approval for all removal tokens

Reply:
hmm, maybe it doesn't. will check

Reply:
we were approving it as an operator previously in the mint function which is where I think this requirement came from

Reply:
all tests pass without it :shrug: any idea why we had it to begin with? This isn't something new I added, just something I shifted the responsibility for